src/snowflake/connector/config_parser/parser.py

```diff
@@ -0,0 +1,190 @@
+#
+# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+#
+
+from __future__ import annotations
+
+import logging
+import os
+from collections.abc import Iterable
+from functools import wraps
+from pathlib import Path
+from typing import Callable, Literal, TypeVar
+
+import tomlkit
+from tomlkit import TOMLDocument
+from tomlkit.items import Table
+
+from ..errors import ConfigParserError, ConfigSourceError
+
+_T = TypeVar("_T")
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ConfigOption:
+    def __init__(
+        self,
+        name: str,
+        _root_parser: ConfigParser,
+        _nest_path: list[str],
+        _type: Callable[[str], _T] | None = None,
+        choices: Iterable[_T] | None = None,
+        env_name: str | None | Literal[False] = None,
+    ) -> None:
+        """Create a config option that can read values from different locations.
+
+        Args:
+            name: The name of the ConfigOption
+            env_name: Environmental variable value should be read from, if not supplied, we'll construct this
+            type: A function that can turn str to the desired type, useful for reading value from environmental variable
```

Comment:
It seems a little bit complicated to me. Do we want to support in our first version? Because for mandatory fields in connection parameters, they are all strings?

Reply:
+1 to what @sfc-gh-jdu . Personally I'm in favour of using methods like `getboolean` or `getfloat` when reading the values from config instead of converting them automatically to particular type.

Reply:
I'd prefer this over that because the `_type` function only needs to be setup once and then we'd always be able to do `parser["option_name"]` to retrieve the values over-and-over again.
With the API you're proposing every usage would need to make sure that they're casting values correctly, so `parser["option_name"]` would become `parser["option_name"].getboolean()` possibly multiple times.

I do want to acknowledge that the name `_type` is not good now weeks after I wrote that piece of code :) It did make a lot of sense that day...

Reply:
I have renamed `_type` to `parse_str`, but otherwise kept the API the same. Hopefully my previous comment makes sense and is convincing